### PR TITLE
Add MapStruct native support

### PR DESCRIPTION
**Pull request by jamesnetherton**

### CODE REVIEW

File: extensions/mapstruct/deployment/src/main/java/org/apache/camel/quarkus/component/mapstruct/deployment/MapStructProcessor.java [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+                     *    public ToCarConversionMethod(CarMapperImpl mapper) {
+                     *        this.mapper = mapper;
+                     *    }
+                     *
+                     *    @Override
+                     *    public Object doConvert(Class<?> type, Exchange exchange, Object value) throws Exception {
+                     *        return mapper.stringToInt(value);
+                     *    }
+                     * }
+                     */
```
Comment (aldettinger) [inline extensions/mapstruct/deployment/src/main/java/org/apache/camel/quarkus/component/mapstruct/deployment/MapStructProcessor.java:206]: :+1:  Nice doc

**Review by oscerd — APPROVED** [review]

File: extensions/mapstruct/runtime/src/main/java/org/apache/camel/quarkus/component/mapstruct/CamelQuarkusMapStructMapperFinder.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    @Override
+    public int discoverMappings(Class<?> clazz) {
+        // Discovery is done at build time so just return the count
+        return mappingsCount;
+    }
+
+    @Override
+    protected void doInit() throws Exception {
+        if (ObjectHelper.isNotEmpty(mapperPackageName)) {
+            LOG.infof("Discovered %d MapStruct type converters from classpath scanning: %s", mappingsCount, mapperPackageName);
```
Comment (aldettinger) [inline extensions/mapstruct/runtime/src/main/java/org/apache/camel/quarkus/component/mapstruct/CamelQuarkusMapStructMapperFinder.java:58]: "from classpath scanning" => "from build time augmentation" ?

File: integration-tests/mapstruct/src/test/java/org/apache/camel/quarkus/component/mapstruct/it/MapStructExplicitPackagesTestProfile.java [hunk 3]
```diff
@@ -16,22 +16,16 @@
...
+import org.apache.camel.quarkus.component.mapstruct.it.mapper.vehicle.VehicleMapper;
 
+public class MapStructExplicitPackagesTestProfile implements QuarkusTestProfile {
     @Override
-    public void configure() throws Exception {
-        from("direct:component").to("mapstruct:" + Car.class.getName());
-        from("direct:converter").convertBodyTo(Car.class);
+    public Map<String, String> getConfigOverrides() {
+        return Map.of("camel.component.mapstruct.mapper-package-name",
+                String.join(",", CarMapper.class.getPackageName(), VehicleMapper.class.getPackageName()));
```
Comment (aldettinger) [inline integration-tests/mapstruct/src/test/java/org/apache/camel/quarkus/component/mapstruct/it/MapStructExplicitPackagesTestProfile.java:29]: Are extra space supported in the list ? like below:
String.join(",        ", CarMapper.class.getPackageName(), VehicleMapper.class.getPackageName()));

Where are they trimmed ?

**Comment by aldettinger** [comment]
From what I understand, config `camel.component.mapstruct.mapStructConverter` is not special enough to deserve a CQ test.

This extension really take advantage of build time augmentation, nice job @jamesnetherton  :+1:

**Review by aldettinger — APPROVED** [review]
Few nitpick if time, beyond it's good.

File: extensions/mapstruct/deployment/src/main/java/org/apache/camel/quarkus/component/mapstruct/deployment/MapStructProcessor.java (hunk 1 above)
Comment (jamesnetherton) [reply extensions/mapstruct/deployment/src/main/java/org/apache/camel/quarkus/component/mapstruct/deployment/MapStructProcessor.java:206]: Just noticed a few copy / paste errors that make it confusing (constructor method names) so I will correct that.

File: extensions/mapstruct/runtime/src/main/java/org/apache/camel/quarkus/component/mapstruct/CamelQuarkusMapStructMapperFinder.java (hunk 2 above)
Comment (jamesnetherton) [reply extensions/mapstruct/runtime/src/main/java/org/apache/camel/quarkus/component/mapstruct/CamelQuarkusMapStructMapperFinder.java:58]: Yeah, I can change it. It was a copy / paste from the default impl that Camel uses.

File: integration-tests/mapstruct/src/test/java/org/apache/camel/quarkus/component/mapstruct/it/MapStructExplicitPackagesTestProfile.java (hunk 3 above)
Comment (jamesnetherton) [reply integration-tests/mapstruct/src/test/java/org/apache/camel/quarkus/component/mapstruct/it/MapStructExplicitPackagesTestProfile.java:29]: Good point. I will fix that.

File: extensions/mapstruct/runtime/src/main/java/org/apache/camel/quarkus/component/mapstruct/CamelQuarkusMapStructMapperFinder.java (hunk 2 above)
Comment (jamesnetherton) [reply extensions/mapstruct/runtime/src/main/java/org/apache/camel/quarkus/component/mapstruct/CamelQuarkusMapStructMapperFinder.java]: Updated the message accordingly.

File: integration-tests/mapstruct/src/test/java/org/apache/camel/quarkus/component/mapstruct/it/MapStructExplicitPackagesTestProfile.java (hunk 3 above)
Comment (jamesnetherton) [reply integration-tests/mapstruct/src/test/java/org/apache/camel/quarkus/component/mapstruct/it/MapStructExplicitPackagesTestProfile.java]: Should be fixed in `MapStructProcessor.getMapperPackages`. I used a helper method from `commons-lang3`. It's a transitive from `quarkus-core-deployment` so I think it's safe.

I also tweaked the `TestProfile` to include whitespace in the config prop as a test for it.

**Comment by jamesnetherton** [comment]
Thanks @aldettinger. Latest commit should address your review comments.

**Pull request merged**